VerCors crash upon returned Viper error

The following input: 
```java
public class C {
    final int[] arr;

    /*@
        inline resource common() =
            arr != null ** Value(arr[*]) **
            (\forall int i = 0 .. arr.length; {:arr[i]:} != 0) ** 
            (\forall int i = 0 .. arr.length, int j = 0 .. i; {:arr[i]:} != {:arr[j]:})
        ;
    @*/

    /*@
    context common();
    requires 0 <= i && i < arr.length;
  @*/
    public /*@ pure @*/ int get(int i) {
        return arr[i];
    }
}
```
Gives the following output:
```
[INFO] Starting verification
[ERROR] Viper returned an error that VerCors does not recognize: Interaction with prover (instance 12) failed: Unexpected output of prover. Expected 'success' but found: (error "line 1082 column 30: unknown constant j@43@00") (<no position>)
```

Using the option --backend-file temp.vpr, I get a viper file which gives the following error when verifying: ```Consistency error: Function post-conditions must not contain permissions. (temp.vpr@106.12--106.197)```.

Removing either of the \forall statements removes this error and causes a successful verification, though the consistency error when verifying temp.vpr remains.

Changing ```context common();``` to ```requires common();``` removes the crash in VerCors and the error when verifying temp.vpr.

--- a/src/col/vct/col/ast/Node.scala
+++ b/src/col/vct/col/ast/Node.scala
@@ -1312,7 +1312,7 @@ sealed trait Expr[G] extends NodeFamily[G] with ExprImpl[G]
 // For terms which we can safely reason about in ExpressionEqualityCheck
 sealed trait SymbolicTerm[G] extends Expr[G]
 // For terms which introduce a resource in a function precondition, invocations of these functions must be excluded from SymbolicTerms that are collected
-sealed trait ResourceTerm[G] extends Expr[G]
+sealed trait ResourceTerm[G] extends Expr[G] with ResourceTermImpl[G]
 
 sealed trait Constant[G] extends Expr[G] with ConstantImpl[G]
 sealed trait ConstantInt[G] extends Constant[G] with ConstantIntImpl[G]

--- a/src/col/vct/col/ast/declaration/category/AbstractFunctionImpl.scala
+++ b/src/col/vct/col/ast/declaration/category/AbstractFunctionImpl.scala
@@ -6,6 +6,8 @@ import vct.col.origin.{Blame, ContractedFailure}
 
 trait AbstractFunctionImpl[G] extends ContractApplicableImpl[G] {
   this: AbstractFunction[G] =>
+
+  override def pure: Boolean = false
   override def body: Option[Expr[G]]
   def threadLocal: Boolean
   override def check(context: CheckContext[G]): Seq[CheckError] =

--- a/src/col/vct/col/ast/declaration/category/ContractApplicableImpl.scala
+++ b/src/col/vct/col/ast/declaration/category/ContractApplicableImpl.scala
@@ -1,18 +1,19 @@
 package vct.col.ast.declaration.category
 
-import vct.col.ast.util.Declarator
 import vct.col.ast.{
   ApplicableContract,
   ContractApplicable,
   Declaration,
   Variable,
 }
-import vct.col.origin.{Blame, ContractedFailure, PostconditionFailed}
+import vct.col.origin.{Blame, ContractedFailure}
 
 trait ContractApplicableImpl[G] extends InlineableApplicableImpl[G] {
   this: ContractApplicable[G] =>
   def contract: ApplicableContract[G]
   def blame: Blame[ContractedFailure]
+  def pure: Boolean
+
   override def declarations: Seq[Declaration[G]] =
     super.declarations ++ contract.givenArgs ++ contract.yieldsArgs ++ typeArgs
 

--- a/src/col/vct/col/ast/expr/apply/ApplyImpl.scala
+++ b/src/col/vct/col/ast/expr/apply/ApplyImpl.scala
@@ -1,9 +1,10 @@
 package vct.col.ast.expr.apply
 
+import vct.col.ast.node.NodeFamilyImpl
 import vct.col.ast.{Applicable, Apply, Expr, Type}
 import vct.col.ref.Ref
 
-trait ApplyImpl[G] {
+trait ApplyImpl[G] extends NodeFamilyImpl[G] {
   this: Apply[G] =>
   def ref: Ref[G, _ <: Applicable[G]]
   def args: Seq[Expr[G]]

--- a/src/col/vct/col/ast/expr/apply/InvocationImpl.scala
+++ b/src/col/vct/col/ast/expr/apply/InvocationImpl.scala
@@ -1,9 +1,30 @@
 package vct.col.ast.expr.apply
 
-import vct.col.ast.{Invocation, Type, Variable}
+import vct.col.ast.{Applicable, Invocation, Local, Type, Variable}
+import vct.col.check.{
+  CheckContext,
+  CheckError,
+  RecursiveFunctionWithoutTerminationMeasure,
+}
+import vct.col.ref.Ref
 
 trait InvocationImpl[G] extends ApplyImpl[G] {
   this: Invocation[G] =>
   override def t: Type[G] = super.t.particularize(typeEnv)
   def typeEnv: Map[Variable[G], Type[G]]
+
+  override def check(context: CheckContext[G]): Seq[CheckError] = {
+    if (
+      context.currentApplicable
+        .contains(ref.decl.asInstanceOf[Applicable[G]]) && ref.decl.pure &&
+      (context.inPreCondition || context.inPostCondition) &&
+      ref.decl.contract.decreases.isEmpty
+    ) {
+      val suggestResult = args.zip(ref.decl.args).forall {
+        case (Local(Ref(a)), p) => a == p
+        case _ => false
+      }
+      Seq(RecursiveFunctionWithoutTerminationMeasure(this, suggestResult))
+    } else { Nil }
+  }
 }

--- a/src/col/vct/col/ast/expr/apply/InvocationImpl.scala
+++ b/src/col/vct/col/ast/expr/apply/InvocationImpl.scala
@@ -4,6 +4,7 @@ import vct.col.ast.{Applicable, Invocation, Local, Type, Variable}
 import vct.col.check.{
   CheckContext,
   CheckError,
+  IncorrectArgumentAmount,
   RecursiveFunctionWithoutTerminationMeasure,
 }
 import vct.col.ref.Ref
@@ -14,6 +15,7 @@ trait InvocationImpl[G] extends ApplyImpl[G] {
   def typeEnv: Map[Variable[G], Type[G]]
 
   override def check(context: CheckContext[G]): Seq[CheckError] = {
+    var result = super.check(context)
     if (
       context.currentApplicable
         .contains(ref.decl.asInstanceOf[Applicable[G]]) && ref.decl.pure &&
@@ -24,7 +26,15 @@ trait InvocationImpl[G] extends ApplyImpl[G] {
         case (Local(Ref(a)), p) => a == p
         case _ => false
       }
-      Seq(RecursiveFunctionWithoutTerminationMeasure(this, suggestResult))
-    } else { Nil }
+      result =
+        result :+
+          RecursiveFunctionWithoutTerminationMeasure(this, suggestResult)
+    }
+    if (ref.decl.args.length != args.length) {
+      result =
+        result :+
+          IncorrectArgumentAmount(this, args.length, ref.decl.args.length)
+    }
+    result
   }
 }

--- a/src/col/vct/col/ast/expr/misc/OldImpl.scala
+++ b/src/col/vct/col/ast/expr/misc/OldImpl.scala
@@ -1,11 +1,17 @@
 package vct.col.ast.expr.misc
 
-import vct.col.ast.{Old, Type}
+import vct.col.ast.node.NodeFamilyImpl
+import vct.col.ast.{ContractApplicable, Old, Type}
 import vct.col.print.{Ctx, Doc, Precedence, Text}
 import vct.col.ast.ops.OldOps
-import vct.col.check.{CheckContext, CheckError, OldInPrecondition}
+import vct.col.check.{
+  CheckContext,
+  CheckError,
+  OldInFunctionContract,
+  OldInPrecondition,
+}
 
-trait OldImpl[G] extends OldOps[G] {
+trait OldImpl[G] extends OldOps[G] with NodeFamilyImpl[G] {
   this: Old[G] =>
   override def t: Type[G] = expr.t
 
@@ -20,9 +26,18 @@ trait OldImpl[G] extends OldOps[G] {
         Text("\\old[") <> ctx.name(at) <> "](" <> expr <> ")"
     }
 
-  override def check(context: CheckContext[G]): Seq[CheckError] =
-    if (context.inPreCondition)
-      Seq(OldInPrecondition(this))
-    else
-      Nil
+  override def check(context: CheckContext[G]): Seq[CheckError] = {
+    var result = super.check(context)
+    if (
+      context.currentApplicable.exists {
+        case a: ContractApplicable[G] => a.pure
+        case _ => false
+      }
+    )
+      result = result :+ OldInFunctionContract(this)
+    else if (context.inPreCondition)
+      result = result :+ OldInPrecondition(this)
+
+    result
+  }
 }

--- a/src/col/vct/col/ast/expr/resource/ResourceTermImpl.scala
+++ b/src/col/vct/col/ast/expr/resource/ResourceTermImpl.scala
@@ -1,0 +1,18 @@
+package vct.col.ast.expr.resource
+
+import vct.col.ast.{ContractApplicable, ResourceTerm}
+import vct.col.ast.node.NodeImpl
+import vct.col.check.{CheckContext, CheckError, ResourceInPostcondition}
+
+trait ResourceTermImpl[G] extends NodeImpl[G] {
+  this: ResourceTerm[G] =>
+
+  override def check(context: CheckContext[G]): Seq[CheckError] =
+    if (
+      context.inPreCondition &&
+      context.currentApplicable.exists { case a: ContractApplicable[G] =>
+        a.pure
+      }
+    ) { Seq(ResourceInPostcondition(this)) }
+    else { Nil }
+}

--- a/src/col/vct/col/ast/expr/resource/ResourceTermImpl.scala
+++ b/src/col/vct/col/ast/expr/resource/ResourceTermImpl.scala
@@ -9,9 +9,9 @@ trait ResourceTermImpl[G] extends NodeImpl[G] {
 
   override def check(context: CheckContext[G]): Seq[CheckError] =
     if (
-      context.inPostCondition &&
-      context.currentApplicable.exists { case a: ContractApplicable[G] =>
-        a.pure
+      context.inPostCondition && context.currentApplicable.exists {
+        case a: ContractApplicable[G] => a.pure
+        case _ => false
       }
     ) { Seq(ResourceInPostcondition(this)) }
     else { Nil }

--- a/src/col/vct/col/ast/expr/resource/ResourceTermImpl.scala
+++ b/src/col/vct/col/ast/expr/resource/ResourceTermImpl.scala
@@ -9,7 +9,7 @@ trait ResourceTermImpl[G] extends NodeImpl[G] {
 
   override def check(context: CheckContext[G]): Seq[CheckError] =
     if (
-      context.inPreCondition &&
+      context.inPostCondition &&
       context.currentApplicable.exists { case a: ContractApplicable[G] =>
         a.pure
       }

--- a/src/col/vct/col/check/Check.scala
+++ b/src/col/vct/col/check/Check.scala
@@ -171,11 +171,19 @@ sealed trait CheckError {
             s"This class cannot extend or implement the type $support since it is not a class"
         )
       case OldInPrecondition(expr) =>
-        Seq(context(expr) -> s"\\old may not be used in a precondition")
+        Seq(context(expr) -> "\\old may not be used in a precondition")
       case ResourceInPostcondition(expr) =>
         Seq(
           context(expr) ->
-            s"Resource terms may not appear in the postcondition of functions (a.k.a pure procedures)"
+            "Resource terms may not appear in the postcondition of functions (a.k.a pure procedures)"
+        )
+      case RecursiveFunctionWithoutTerminationMeasure(expr, suggestResult) =>
+        Seq(
+          context(expr) ->
+            ("Recursive function calls in contracts are only allowed for functions with a termination measure" +
+              (if (suggestResult) {
+                 " (Hint: use \\result to refer to the output of the function)"
+               } else { "" }))
         )
     }): _*)
 
@@ -302,6 +310,12 @@ case class OldInPrecondition(expr: Node[_]) extends CheckError {
 }
 case class ResourceInPostcondition(node: Node[_]) extends CheckError {
   val subcode = "resourceInPostcondition"
+}
+case class RecursiveFunctionWithoutTerminationMeasure(
+    node: Node[_],
+    suggestResult: Boolean,
+) extends CheckError {
+  val subcode = "missingTerminationMeasure"
 }
 
 case object CheckContext {

--- a/src/col/vct/col/check/Check.scala
+++ b/src/col/vct/col/check/Check.scala
@@ -172,6 +172,11 @@ sealed trait CheckError {
         )
       case OldInPrecondition(expr) =>
         Seq(context(expr) -> s"\\old may not be used in a precondition")
+      case ResourceInPostcondition(expr) =>
+        Seq(
+          context(expr) ->
+            s"Resource terms may not appear in the postcondition of functions (a.k.a pure procedures)"
+        )
     }): _*)
 
   def subcode: String
@@ -294,6 +299,9 @@ case class SupportNotAClass(cls: Node[_], support: Type[_]) extends CheckError {
 }
 case class OldInPrecondition(expr: Node[_]) extends CheckError {
   val subcode = "oldInPrecondition"
+}
+case class ResourceInPostcondition(node: Node[_]) extends CheckError {
+  val subcode = "resourceInPostcondition"
 }
 
 case object CheckContext {

--- a/src/col/vct/col/check/Check.scala
+++ b/src/col/vct/col/check/Check.scala
@@ -172,6 +172,11 @@ sealed trait CheckError {
         )
       case OldInPrecondition(expr) =>
         Seq(context(expr) -> "\\old may not be used in a precondition")
+      case OldInFunctionContract(expr) =>
+        Seq(
+          context(expr) ->
+            "\\old may not be used in function (a.k.a pure procedure) contracts"
+        )
       case ResourceInPostcondition(expr) =>
         Seq(
           context(expr) ->
@@ -312,6 +317,9 @@ case class SupportNotAClass(cls: Node[_], support: Type[_]) extends CheckError {
 }
 case class OldInPrecondition(expr: Node[_]) extends CheckError {
   val subcode = "oldInPrecondition"
+}
+case class OldInFunctionContract(expr: Node[_]) extends CheckError {
+  val subcode = "oldInFunction"
 }
 case class ResourceInPostcondition(node: Node[_]) extends CheckError {
   val subcode = "resourceInPostcondition"

--- a/src/col/vct/col/check/Check.scala
+++ b/src/col/vct/col/check/Check.scala
@@ -185,6 +185,11 @@ sealed trait CheckError {
                  " (Hint: use \\result to refer to the output of the function)"
                } else { "" }))
         )
+      case IncorrectArgumentAmount(expr, gotCount, expectedCount) =>
+        Seq(
+          context(expr) ->
+            s"This invocation has the wrong number of arguments, got: $gotCount expected: $expectedCount"
+        )
     }): _*)
 
   def subcode: String
@@ -316,6 +321,14 @@ case class RecursiveFunctionWithoutTerminationMeasure(
     suggestResult: Boolean,
 ) extends CheckError {
   val subcode = "missingTerminationMeasure"
+}
+// Mostly for catching wrong generated calls (Resolution catches most of these when they come from the user)
+case class IncorrectArgumentAmount(
+    node: Node[_],
+    gotCount: Int,
+    expectedCount: Int,
+) extends CheckError {
+  val subcode = "incorrectArgumentAmount"
 }
 
 case object CheckContext {

--- a/src/rewrite/vct/rewrite/ResolveExpressionSideEffects.scala
+++ b/src/rewrite/vct/rewrite/ResolveExpressionSideEffects.scala
@@ -592,7 +592,7 @@ case class ResolveExpressionSideEffects[Pre <: Generation]()
         stored(value, oldValue.t)
       case ass @ PostAssignExpression(oldTarget, value) =>
         val target = assignTarget(oldTarget)
-        val cachedTarget = stored(target, oldTarget.t)
+        val cachedTarget = notInlined(oldTarget)
         effect(Assign(target, inlined(value))(ass.blame)(e.o))
         stored(cachedTarget, oldTarget.t)
       case With(pre, value) =>

--- a/src/rewrite/vct/rewrite/ResolveExpressionSideEffects.scala
+++ b/src/rewrite/vct/rewrite/ResolveExpressionSideEffects.scala
@@ -591,8 +591,8 @@ case class ResolveExpressionSideEffects[Pre <: Generation]()
         effect(Assign(target, value)(ass.blame)(e.o))
         stored(value, oldValue.t)
       case ass @ PostAssignExpression(oldTarget, value) =>
+        val cachedTarget = stored(inlined(oldTarget), oldTarget.t)
         val target = assignTarget(oldTarget)
-        val cachedTarget = notInlined(oldTarget)
         effect(Assign(target, inlined(value))(ass.blame)(e.o))
         stored(cachedTarget, oldTarget.t)
       case With(pre, value) =>

--- a/test/main/vct/test/integration/examples/QualifierSpec.scala
+++ b/test/main/vct/test/integration/examples/QualifierSpec.scala
@@ -1049,7 +1049,7 @@ int get_first_two(/*@unique<1>@*/ int* xs){
   vercors should error withCode "disallowedQualifiedMethodCoercionNest" in "Method has func in contract - 2" c
     """
  /*@
-  context xs != NULL ** \pointer_length(xs)>2 ** Perm(xs[0], 1\100);
+  requires xs != NULL ** \pointer_length(xs)>2 ** Perm(xs[0], 1\100);
   ensures \result == xs[0];
 @*/
 /*@ pure @*/ int id(/*@unique<2>@*/ int* xs){

